File: packages/domain/src/adapters/registry.ts
/**
 * In-memory source registry with register/get/list/approve policy.
 * Firestore persistence implements SourceRegistryStore in a later.
 */
import { assertEvidenceSourceValid } from '../provenance/source.js';
import { assertSourceAdapterContractValid } from './contract.js';
import type { AdapterRegistryState, SourceAdapterContract, SourceRegistryEntry } from './types.js';

export type RegisterSourceInput = {
  readonly id: string;
  readonly contract: SourceAdapterContract;
  readonly evidenceSource: SourceRegistryEntry['evidenceSource'];
  readonly registryState?: AdapterRegistryState;
  readonly createdAt: string;
};

export type ApproveSourcePolicyInput = {
  readonly id: string;
  readonly approvedBy: string;
  readonly approvedAt: string;
  readonly registryState?: Extract<AdapterRegistryState, 'approved' | 'canary'>;
};

/** Persistence boundary for registry entries (Firestore adapter implements later). */
export type SourceRegistryStore = {
  get(id: string): SourceRegistryEntry | undefined;
  list(): readonly SourceRegistryEntry[];
  save(entry: SourceRegistryEntry): void;
};

export function createInMemorySourceRegistry(
  seed: readonly SourceRegistryEntry[] = [],
): SourceRegistryStore {
  const entries = new Map<string, SourceRegistryEntry>(seed.map((entry) => [entry.id, entry]));
  return {
    get(id: string) {
      return entries.get(id);
    },
    list() {
      return [...entries.values()];
    },
    save(entry: SourceRegistryEntry) {
      entries.set(entry.id, entry);
    },
  };
}

export function registerSource(
  store: SourceRegistryStore,
  input: RegisterSourceInput,
): SourceRegistryEntry {
  assertSourceAdapterContractValid(input.contract);
  assertEvidenceSourceValid(input.evidenceSource);
  if (input.evidenceSource.adapterId !== input.contract.adapterId) {
    throw new Error(
      `evidenceSource.adapterId (${input.evidenceSource.adapterId}) must match contract.adapterId (${input.contract.adapterId})`,
    );
  }
  if (store.get(input.id)) {
    throw new Error(`Source registry entry already exists: ${input.id}`);
  }
  const entry: SourceRegistryEntry = {
    id: input.id,
    contract: input.contract,
    evidenceSource: input.evidenceSource,
    registryState: input.registryState ?? 'disabled',
    createdAt: input.createdAt,
    updatedAt: input.createdAt,
  };
  store.save(entry);
  return entry;
}

export function getSourceEntry(store: SourceRegistryStore, id: string): SourceRegistryEntry | undefined {
  return store.get(id);
}

export function listSourceEntries(
  store: SourceRegistryStore,
  filter?: { readonly registryState?: AdapterRegistryState },
): readonly SourceRegistryEntry[] {
  const all = store.list();
  if (!filter?.registryState) {
    return all;
  }
  return all.filter((entry) => entry.registryState === filter.registryState);
}

export function approveSourcePolicy(
  store: SourceRegistryStore,
  input: ApproveSourcePolicyInput,
): SourceRegistryEntry {
  const existing = store.get(input.id);
  if (!existing) {
    throw new Error(`Source registry entry not found: ${input.id}`);
  }
  if (existing.registryState === 'quarantined' || existing.registryState === 'dead_letter') {
    throw new Error(
      `Cannot approve source policy in state "${existing.registryState}"; resolve quarantine/dead-letter first`,
    );
  }
  const updated: SourceRegistryEntry = {
    ...existing,
    registryState: input.registryState ?? 'approved',
    approvedAt: input.approvedAt,
    approvedBy: input.approvedBy,
    updatedAt: input.approvedAt,
  };
  store.save(updated);
  return updated;
}

export function setRegistryState(
  store: SourceRegistryStore,
  input: {
    readonly id: string;
    readonly registryState: AdapterRegistryState;
    readonly updatedAt: string;
    readonly quarantineReason?: string;
    readonly deadLetterReason?: string;
  },
): SourceRegistryEntry {
  const existing = store.get(input.id);
  if (!existing) {
    throw new Error(`Source registry entry not found: ${input.id}`);
  }
  const updated: SourceRegistryEntry = {
    ...existing,
    registryState: input.registryState,
    updatedAt: input.updatedAt,
    ...(input.quarantineReason ? { quarantineReason: input.quarantineReason } : {}),
    ...(input.deadLetterReason ? { deadLetterReason: input.deadLetterReason } : {}),
  };
  store.save(updated);
  return updated;
}

File: workers/research/src/black_book_research/adapters/registry.py
"""In-memory source registry mirroring domain register/get/list/approve (BB-037)."""

from __future__ import annotations

from dataclasses import replace

from .types import (
    AdapterRegistryState,
    EvidenceSource,
    SourceAdapterContract,
    SourceRegistryEntry,
    SourceRegistryStore,
)


class InMemorySourceRegistry:
    def __init__(self, seed: list[SourceRegistryEntry] | None = None) -> None:
        self._entries: dict[str, SourceRegistryEntry] = {
            entry.id: entry for entry in (seed or [])
        }

    def get(self, entry_id: str) -> SourceRegistryEntry | None:
        return self._entries.get(entry_id)

    def list(self) -> list[SourceRegistryEntry]:
        return list(self._entries.values())

    def save(self, entry: SourceRegistryEntry) -> None:
        self._entries[entry.id] = entry


def create_in_memory_source_registry(
    seed: list[SourceRegistryEntry] | None = None,
) -> SourceRegistryStore:
    return InMemorySourceRegistry(seed)


def register_source(
    store: SourceRegistryStore,
    *,
    entry_id: str,
    contract: SourceAdapterContract,
    evidence_source: EvidenceSource,
    created_at: str,
    registry_state: AdapterRegistryState = "disabled",
) -> SourceRegistryEntry:
    if store.get(entry_id):
        raise ValueError(f"Source registry entry already exists: {entry_id}")
    if evidence_source.adapter_id != contract.adapter_id:
        raise ValueError("evidence_source.adapter_id must match contract.adapter_id")
    entry = SourceRegistryEntry(
        id=entry_id,
        contract=contract,
        evidence_source=evidence_source,
        registry_state=registry_state,
        created_at=created_at,
        updated_at=created_at,
    )
    store.save(entry)
    return entry


def list_source_entries(
    store: SourceRegistryStore,
    *,
    registry_state: AdapterRegistryState | None = None,
) -> list[SourceRegistryEntry]:
    entries = store.list()
    if registry_state is None:
        return entries
    return [entry for entry in entries if entry.registry_state == registry_state]


def approve_source_policy(
    store: SourceRegistryStore,
    *,
    entry_id: str,
    approved_by: str,
    approved_at: str,
    registry_state: AdapterRegistryState = "approved",
) -> SourceRegistryEntry:
    existing = store.get(entry_id)
    if existing is None:
        raise ValueError(f"Source registry entry not found: {entry_id}")
    if existing.registry_state in {"quarantined", "dead_letter"}:
        raise ValueError(
            f"Cannot approve source policy in state {existing.registry_state!r}"
        )
    updated = replace(
        existing,
        registry_state=registry_state,
        approved_at=approved_at,
        approved_by=approved_by,
        updated_at=approved_at,
    )
    store.save(updated)
    return updated


def set_registry_state(
    store: SourceRegistryStore,
    *,
    entry_id: str,
    registry_state: AdapterRegistryState,
    updated_at: str,
    quarantine_reason: str | None = None,
    dead_letter_reason: str | None = None,
) -> SourceRegistryEntry:
    existing = store.get(entry_id)
    if existing is None:
        raise ValueError(f"Source registry entry not found: {entry_id}")
    updated = replace(
        existing,
        registry_state=registry_state,
        updated_at=updated_at,
        quarantine_reason=quarantine_reason,
        dead_letter_reason=dead_letter_reason,
    )
    store.save(updated)
    return updated
